### PR TITLE
testing for PNG and JPEG

### LINKED ISSUE #168: macos confusion about libpbg version
When I added a test for PNG and it ran on the mac, I get this:

```
12/12 Test #12: tst_png ..........................***Failed    0.10 sec
33
libpng warning: Application built with libpng-1.4.12 but running with 1.6.37
34
Testing PNG functions.
35
Testing enc_png() call...-1
```

Turns out this is a known macos issue: https://stackoverflow.com/questions/36523911/osx-homebrew-cmake-libpng-version-mismatch-issue

### PR DESCRIPTION
**Pull request by edwardhartnett**
Part of #8.
Fixes #168 
Part of #170.

More testing, this time for encoding/decoding with png and jpeg.

Also lots of code cleanup, and doxygen improvements.

### CODE REVIEW

**Comment by edwardhartnett** [comment]
OK, this takes us up to ~39% code coverage - certainly an improvement. There are still many packing/unpacking code files with no testing.

image

**Comment by edwardhartnett** [comment]
@Hang-Lei-NOAA and @BrianCurtis-NOAA please review so I can commit and continue in a new PR...

File: src/dec_jpeg2000.c [hunk 1]
```diff
@@ -36,96 +36,83 @@ void dummy(void) {}
...
 
-    jpcstream=jas_stream_memopen(injpc,bufsize);
-
-//
-//     Decode JPEG200 codestream into jas_image_t structure.
-//
-    image=jpc_decode(jpcstream,opts);
+    /* Decode JPEG200 codestream into jas_image_t structure. */
+    image=jpc_decode(jpcstream, opts);
     if ( image == 0 ) {
```
Comment (BrianCurtis-NOAA) [inline src/dec_jpeg2000.c]: should this be `if ( !image )` ?

Comment (edwardhartnett) [reply src/dec_jpeg2000.c]: OK...

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Comment by edwardhartnett** [comment]
Up to 42%!

image

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,9 +34,12 @@ if(NOT CMAKE_BUILD_TYPE MATCHES "^(Debug|Release|RelWithDebInfo|MinSizeRel)$")
                                                "MinSizeRel" "RelWithDebInfo")
 endif()
 
+link_directories("/usr/local/lib")
 if(USE_PNG)
   find_package(ZLIB REQUIRED)
-  find_package(PNG REQUIRED)
+  find_package(PkgConfig REQUIRED)
+  pkg_check_modules(PNG libpng16 REQUIRED)
+#  find_package(PNG REQUIRED)
```
Comment (DusanJovic-NOAA) [inline CMakeLists.txt:42]: Why was this necessary? Are there some issues with find_package? @aerorahul

Comment (edwardhartnett) [reply CMakeLists.txt:42]: In order to get the PNG build to work on the Mac.

Comment (DusanJovic-NOAA) [reply CMakeLists.txt:42]: Then the changes required to fix broken PNG installation on Mac should be in:
```
if(APPLE)
 ...
endif()
```

Comment (aerorahul) [reply CMakeLists.txt:42]: I had no issues with `find_package` on my macOS linking w/ PNG.
I can still build with this:
```
#link_directories("/usr/local/lib")
if(USE_PNG)
  find_package(ZLIB REQUIRED)
  #find_package(PkgConfig REQUIRED)
  #pkg_check_modules(PNG libpng16 REQUIRED)
  find_package(PNG REQUIRED)
else()
  message(STATUS "Will not build PNG support")
endif()
```

The library builds and all test pass.

Comment (aerorahul) [reply CMakeLists.txt:42]: The issue is not really w/ macOS.  Its got to do w/ how the Github Action macOS is configured.
`brew` is used to install `libpng`, which is likely the issue that causes the action to fail.
Like, `jasper`, if PNG is built and cached, I am sure this won't be an issue and we can revert to "standard operating procedures"
I don't have time to investigate further at the moment.

**Comment by edwardhartnett** [comment]
OK, good idea @aerorahul I will try to build PNG from source and cache it.

**Comment by DusanJovic-NOAA** [comment]
How is that build going to be different than brew's build?

**Pull request merged**